File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left,maxlen=0,0
        st=set()
        for i in range(len(s)):
            while s[i] in st:
                st.remove(s[left])
                left+=1
            st.add(s[i])
            maxlen=max(maxlen,i-left+1)
        return maxlen        


                   
        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> st=new HashSet<>();
        int left=0,maxlen=0;
        for(int i=0;i<s.length();i++){
            while(st.contains(s.charAt(i))){
                st.remove(s.charAt(left));
                left++;
            }
            st.add(s.charAt(i));
            maxlen=Math.max(maxlen,i-left+1);
        }
        return maxlen;
    }
}